signac/core/collection_json.py

```diff
@@ -11,10 +11,73 @@
 import json
 import errno
 import uuid
+import warnings
 
 from .synced_collection import SyncedCollection
 from .syncedattrdict import SyncedAttrDict
 from .synced_list import SyncedList
+from collections.abc import Mapping
+from collections.abc import Sequence
+from .synced_collection import NUMPY
+from ..errors import KeyTypeError
+
+if NUMPY:
+    import numpy
+
+
+class JSONFormatValidator:
+    """Implement the validation for :class:`JSONCollection`."""
+
+    @classmethod
+    def validate(cls, data):
```

Comment:
@bdice now that the validators are just arbitrary callables and not classes, I think we should perhaps move

Reply:
I think we can remove converting NumPy arrays to lists from validators as `SyncedList` handle NumPy arrays. I moved that to validation because I think the serializing of arrays is done to store to JSON backend. In the current implementation, we handle (b) internally so can be removed from validation.

Reply:
Done.

Reply:
@vyasr @vishav1771 I am in favor of **1) change validators to be either raise exceptions or be no-ops.** We shouldn't make forward-thinking design decisions based on deprecated behavior. As noted above, that would require us to break deprecated behavior if we just dropped the non-str keys. I would consider inserting a "workaround" in `JSONCollection`, which would be separate from (perhaps before?) validation. Its implementation would be similar to the previous approach to handling this. There should be a clear comment to remove the workaround in signac 2.0, since it only affects the JSON backend with specific key types.